### ISSUE TITLE
Hard coded MD5 hashing algorithm prevents FIPS compliance

### ISSUE THREAD

**Issue by woot4moo** [openshift/origin#9045]
[provide a description of the issue]
  Files such as:
https://github.com/openshift/origin/blob/388478c40e751c4295dcb9a44dd69e5ac65d0e3b/pkg/auth/authenticator/password/htpasswd/md5.go

  Rely on hard coded MD5 hashing algorithms.  This results in systems that are FIPS enabled to cause runtime errors.  By removing the hard coded MD5 and replacing it with SHA-2 (or any FIPS compliant hashing algorithm) systems requiring FIPS support can run OpenShift.

##### Version
[provide output of the `openshift version` or `oc version` command]
  master branch
##### Steps To Reproduce
1. [step 1]
2. [step 2]

##### Additional Information
[try to run `$ oadm diagnostics` command if possible]
[if you are reporting issue related to builds, provide build logs with `BUILD_LOGLEVEL=5`]
[consider attaching output of the `$ oc get all -o json -n <namespace>` command to the issue]
[visit https://docs.openshift.org/latest/welcome/index.html]

**Comment by liggitt** [comment]
Unfortunately, the algorithm used for that auth method is not determined by us, but by the apache format.

**Comment by liggitt** [comment]
Can you give more details on the runtime errors caused?

**Comment by woot4moo** [comment]
@liggitt The operating system level error message would look like this:

> 140062425388872:error:060800A3:digital envelope routines:EVP_DigestInit_ex:disabled for fips:digest.c:251

However, I do not understand what you mean by 'the apache format' as I would be hard pressed to believe that MD5 is required

**Comment by liggitt** [comment]
https://httpd.apache.org/docs/current/programs/htpasswd.html

> -m
> Use MD5 encryption for passwords. This is the default (since version 2.2.18).

**Comment by woot4moo** [comment]
Sure, that is the default so it therefore can be changed correct?  Such as the page that further states:

> htpasswd encrypts passwords using either bcrypt, a version of MD5 modified for Apache, SHA1, or the system's crypt() routine

**Comment by liggitt** [comment]
Right... but we can't claim to support reading the htpasswd format and not support the default hashing algorithm... it means we would not work with a default-generated file

**Comment by woot4moo** [comment]
Again agree, I am not suggesting we completely abandon MD5.  What I am suggesting is that instead of hard coding MD5 (and yes I agree that switching to a hard coded SHA2 is just as problematic) we could make it a variable.  Such that we would not preclude systems that require additional security concerns.
